File: frontend/src/lib/api/progress.ts
import { api } from './api'

export interface ProgressConcept {
  name: string
  mastery: number
}

export interface ProgressTopic {
  topic_id: number
  topic_name: string
  mastery: number
  concepts: ProgressConcept[]
}

export interface UserProgress {
  overall_mastery: number
  topics: ProgressTopic[]
}

export interface ProgressHistoryItem {
  date: string
  concept_id: number
  mastery: number
}

export async function getProgress(): Promise<UserProgress> {
  const response = await api.get<UserProgress>('/api/progress')
  return response.data
}

export async function getProgressHistory(): Promise<ProgressHistoryItem[]> {
  const response = await api.get<ProgressHistoryItem[]>('/api/progress/history')
  return response.data
}

File: backend/app/schemas/progress.py
from pydantic import BaseModel


class ProgressConceptOut(BaseModel):
    name: str
    mastery: float


class ProgressTopicOut(BaseModel):
    topic_name: str
    concepts: list[ProgressConceptOut]


class ProgressResponse(BaseModel):
    topics: list[ProgressTopicOut]


class ProgressHistoryItem(BaseModel):
    date: str
    concept_id: int
    concept_name: str
    mastery: float
    is_correct: bool
